### PR TITLE
feat(gui-app,protocol): hash-first image attachments for the create flow

### PR DESCRIPTION
**Pull request by hdkshingala**
Hash-first image attachments for the create flow: a pasted image is prepared once, stored by content hash, and sent as a digest instead of inline base64. The renderer half, its protocol minors, and the client transport.

## Why

Creating a task with attachments on a remote host was timing out. The RCA is that the create request carried every image inline as base64 inside one unary RPC: a handful of screenshots put a multi-megabyte body through the relay's bulk transport, under the epic lease, on the path the user is waiting behind. The fix is to stop moving bytes on that path at all — prepare and store the image when it is pasted, and let the create carry only its digest.

## Shape

**At paste.** `prepare` bounds the bitmap (≤ 2000 px longest edge, ≤ 3.75 MiB, animation preserved verbatim or refused), stores the result in a content-addressed IndexedDB partition, and rewrites the editor node to hash-only. Quality is stepped before scale, because downscaling costs less legibility than recompression. A format the host's writer will not take stays inline rather than being hashed.

**At submit.** Two independent gates, on two independent lines:

| surface | gate | line |
| --- | --- | --- |
| landing composer, in-epic new-conversation modal | `createAttachmentsByHashSupported` | unary `epic.create` / `epic.createChat` **@1.2** |
| chat composer, edit-and-resend, seeded handoff | `ChatStreamClient.draftBlobBridgeSupported()` | stream `chat.subscribe` **@1.12** |

Both also require the host not to be withholding `drafts.putBlob`. Where a gate is closed the old inline path is taken unchanged, so a new client against an old host behaves exactly as it does today.

The two create surfaces deliberately differ on a refusal. The landing composer refuses a create whose bytes it cannot produce, because no epic exists yet and nothing else can resolve them. The in-epic modal is best-effort on the inline remainder, because a chat document routinely holds hashes addressing the epic's own attachment store — an image copied out of a rendered message, a quote seed — and the host resolves those before it ever looks at the staging tier. Refusing them would break creates that work today.

**Budget.** A per-window 64 MiB partition budget with a per-image ceiling, charged over the union of every holder that can outlive a draft clear: composer drafts, modal patches, initial-chat handoffs, pending and refused sends, cancel restorations, queued items, the open inline edit, the parked steer-conflict confirm, and in-flight repairs. A holder that roots an image also prices it; bare-hash holders that carry no size root without pricing.

**Recovery.** A queued item whose staged bytes went missing is repaired from the window's local copy and the queue resumed — bounded to one repair per item per session, gated on the viewer actually being able to act, and re-checked at fulfilment against the live queue, the mount, a paused-episode counter and eligibility, so a cancel, a manual resume or an unmount mid-upload cannot resume someone else's queue.

## Protocol

`epic.create` / `epic.createChat` **@1.2** add `attachmentsByHash` and a typed refusal. `drafts.putBlob` / `drafts.readBlob` **@1.1**. `drafts.claim` is now `drafts.retract`.

This branch originally carried its own `chat.subscribe@1.11` as a send-side capability signal. While it was in flight, `main` shipped the same capability as **@1.12** with its own client gate, so ours was deleted rather than renumbered — two minors claiming one capability make the gate ambiguous. A long-lived branch does not own an unreleased minor.

## Merge

Merged `origin/main` (50 commits) rather than rebased. The resolution decisions worth knowing:

- Where `main` had shipped a capability first, `main`'s mechanism is the one that survives and this branch's duplicate was deleted — the protocol minor above, and `landing-image-store.ts`, whose hash-addressed store this branch had been building in parallel.
- `main`'s blanket refusal of any hash-only node the client could not inline was removed on the `epic.createChat` surface only, for the reason given above; it rests on a premise that does not hold where the epic already exists, and it contradicts `draft-image-inlining.ts`'s own contract that the client never refuses a send because its replica came up empty.

## Three things inherited, not introduced — reported rather than fixed

All three are byte-identical to `main` — the first two as whole files this branch never touches, the third as the exact block quoted — so changing them here would be unrelated churn in an already-large diff:

- `protocol/src/host/agent/gui/subscribe.ts:1-8` and `:2890-2893` still describe `chat.subscribe@1.11` as the live contract and as "the first line whose client may say `auto`". `@1.12` and `@1.13` were both minted on `main` above it, so these went stale there.
- `clients/gui-app/src/components/chat/__tests__/browser-reference-chips.test.tsx`'s `vi.mock` of the image store omits `ensureMeasuredImageSizes`, so the GC's startup reconcile throws a caught error and the reconcile silently does not run. The export already exists on `main`.
- `clients/gui-app/src/components/epic-canvas/renderers/use-chat-message-actions.ts:683` — in `commit`, `current.targetMessageId !== targetMessageId` is dead. A record's target cannot change without its session changing (`targetMessageId` is written at one site, in the same object literal as `sessionId`, and the only dispatcher mints a fresh id per begin), so the session check six lines below strictly dominates it. Removing it reds nothing in either suite. This branch does add to that file, but the whole guard block — both checks and their comments — is byte-for-byte `main`'s, so the dominance is upstream's structure and not something the merge created. Worth knowing rather than fixing, because `:716`'s pin reads as a retarget test while actually being carried by the session check: deleting the session check on the strength of that pin would delete the thing doing the work, and the pin would stay green.

## Testing

Local, on the merged tree, all green:

| check | result |
| --- | --- |
| `bun run compile` | 6 projects |
| `bun run lint` | 6 projects — oxlint **and** ESLint (the script is `oxlint && eslint`, so a green chain means both ran) |
| `oxfmt --list-different` | clean |
| `clients/gui-app` | **2368 files, 30378 tests** — the whole suite, not a derived run set (see Round 3) |
| `clients/shared` | **3 files, 321 tests** |
| `protocol` | **7 files, 97 tests** |

The run set is every test file this branch touches or whose subject it changes, derived from cumulative porcelain rather than from memory. Each run is checked with a set difference between the files vitest **reported** and the files requested, not on the exit code — see below.

**Two things this branch found about its own verification, both of which had been hiding real defects.**

`vitest` exits **0** when a worker fork dies. One run read `Test Files 81 passed (82)` / `Tests 1402 passed (1406)` / `Errors 1 error` with a zero exit. The missing file — `use-initial-chat-handoff-driver-v3-resend.test.tsx`, ours, 4 cells — had never run in any pass on this branch: its worker climbed past 4.5 GB and died, and since nothing failed, nothing was reported. Cause was a callback constructed inside the `renderHook` render function while the hook lists it in an effect dependency array whose body calls `setState` — unbounded render → effect → setState. The production mounter wraps the same getter in `useCallback`, so this was a harness defect, not a shipping one. A sweep of the class across gui-app (110 candidate sites → 7 hook/dep pairs → 1) found no others; the discriminator is whether every state write in the effect sits behind a ref latch.

gui-app's lint is one `&&` chain, so an oxlint failure means **ESLint never executes**. Three successive red runs were oxlint's, under rule names that read like ESLint's (`eslint(complexity)`, `eslint(no-nested-ternary)`). When oxlint finally went clean, ESLint ran for the first time on this branch and reported two `react-hooks/immutability` errors in production code added here. Both are fixed; the row above is a green chain, which does prove ESLint ran.

**What is not covered.** The queued-drain repair arm has unit coverage on both sides (`use-queued-prompt-blob-repair` 19 cells, plus the host's drain tests) but **has not been exercised against a real host end to end** — paste, pause, delete the staged bytes on disk, resume, watch the repair put them back. That row needs a booted dev desktop and a person at the keyboard; it is written up step by step in the epic's execution log, including the correction that the live gate is `ChatStreamClient.draftBlobBridgeSupported()` at `minor >= 12` and not the `@1.11` this branch originally carried.

---

## Round 3 — `main` deleted a plane this branch was built on

Fifteen commits arrived on `main`, and #1979 ("replace the prompt stash with a composer Drafts control") deleted `prompt-stash-image-preparation.ts`. Ten conflicts; the markers were the small part.

**The collision was a naming accident.** That file had stopped being a prompt-stash module before `main` ever touched it: ticket 1 generalised it *in place* into the shared image preparer every composer paste surface now runs — `PREPARED_IMAGE_POLICY`, `createImagePreparationSession`, the encode ladder, `sniffImageDimensions` — and left the stash's own names beside it as aliases. From `main`'s side it was still the stash's file, so #1979 deleted it whole. Taking either side alone was wrong: `main`'s deletion removes the paste engine this PR exists to add, and ours resurrects a plane the product has retired. The file survives its own module and not its name — renamed to `composer-image-preparation-session.ts`, with the stash policy, its two constants, its session factory and its four type aliases removed.

**The 5 MiB per-image ceiling is unchanged and still earns it.** The writer justifying the larger number was the stash restore, and #1979 did not retire that writer — it *renamed* it (`landing-stash-import.ts` → `landing-image-import.ts`) and kept it, because the same change migrates existing stash entries into start-page drafts and that importer still writes their blobs through `putImage` verbatim. A legacy animation can still be 5 MiB. So the constant keeps its value and its derivation, and the bound it derives from moves to the budget as `LEGACY_STASH_IMAGE_MAX_BYTES` — named for the data rather than for a policy, because that is what it now is.

**Coverage an obvious retarget would have dropped.** The preparation suites drove the engine through the stash policy at twenty-odd sites, and pointing them at `PREPARED_IMAGE_POLICY` loses a distinction. The stash's numbers are kept as `SPLIT_CEILING_TEST_POLICY`, whose `byteCeiling` (975 KiB) is below its `animationCeiling` (5 MiB) — a shape no production policy has, since `PREPARED_IMAGE_POLICY` sets the two equal. That split is the point: it is the only way to drive an animation sized BETWEEN the ceilings, which is what separates "kept verbatim because re-encoding cannot be frame-faithful" from "small enough that nothing would have re-encoded it". Where the ceilings are equal that interval is empty.

_(An earlier revision of this section claimed the verbatim-animation branch would otherwise have had "no coverage at all". That was wrong — `prepared-image-policy-matrix.test.ts` already covers it under the production policy and predates this merge. Caught by the round-3 cold review; the fixture's own docblock now states the narrower, true reason.)_

**A fourth instance of a defect this PR fixes arrived with the merge.** `new-conversation-modal-store.ts` is new on `main` and persists `selection` beside a stripped `content` — the same defect fixed here in three other seams hours earlier, arriving from the other side while that fix was being written. It takes `stripBase64ImageNodesWithSelection` like the rest.

**A break neither side's diff could show.** `main` added `useEpicTitle()` to `new-conversation-modal.tsx` and updated the three test files that mock `@/lib/epic-selectors` for that modal; this branch added a fourth. No conflict, no type error — nothing checks a `vi.mock` factory against the module it replaces — so the merged tree compiled and linted clean while that file died at setup, taking all 12 of its tests with it. This is why the Testing row above is now the **whole** gui-app suite rather than a run set derived from the diff: a derived set cannot contain a file whose subject was changed by the *other* side of the merge. The suite is also the sweep for siblings — a missing export in a full-replacement factory throws identically every time, and a whole-suite run found exactly this one file.

One red is **not** fixed and is not this branch's: `providers-settings-panel.test.tsx` (2 of 91). It is byte-identical to `main` here, there is no diff against `main` anywhere under `settings/`, and it passes 91/91 run alone — contention.

**Two calls flagged rather than made silently.**

- *The ceiling* — kept at `main`'s value. It turned out to be derivable rather than a judgment call once the migration importer was found, so no product decision was taken on someone else's file.
- *The modal's prompt-stash image tier* — the stash source it read is gone, so the tier goes with it. If the Drafts control is meant to carry images into a new conversation, that is a feature to add against the new plane, not a hunk to keep alive against a deleted one.

### Downstream

`traycerai/traycer-internal#5743` pins this repo as a submodule and is **conflicting on the gitlink alone** — every source file there auto-merges. That conflict is unresolvable before this PR lands, because the correct pin is a `main` commit containing both sides. After merging here: re-pin the submodule to the new `main`, merge `origin/development` into that branch, and the conflict is closed.

---

## Post-merge cold review (round 3)

A read-only whole-diff review of the merge found one real hole and three claims of mine that were wrong in the direction that reads as thorough. All closed in-branch.

**Coverage the merge deleted without a conflict.** `prompt-stash-image-preparation-animation.test.ts` was on this branch; main deleted it with the rest of the stash plane; this branch never touched it — so git resolved a **clean delete**. No conflict, no hunk in either diff, and nothing type-checks a test file out of existence. But the module it tested was *rescued*, not deleted, into `composer-image-preparation-session.ts`. Thirteen tests vanished off a module that is now more load-bearing than when they were written, because it runs on every composer paste surface rather than on the stash alone.

Among the lost arms was the one the preparer's own comment names as its reason for existing: a verbatim animation is still decode-validated, *"because being kept verbatim is exactly why nothing downstream would ever catch a payload a real decoder rejects"*. That had **no** coverage at all. It does now — ablating decode-validation for animations reddens exactly those two cases and leaves the other twelve green.

The tell was three fixtures carried through the rename with **zero importers** — exactly the deleted file's imports. Their regaining callers is the completeness check, and it is asserted rather than assumed. Swept by predicate: of the test files main deleted that this branch never touched, exactly one targets a module the merge rescued.

**Three claims corrected.** The `SPLIT_CEILING_TEST_POLICY` justification said retargeting would have left the verbatim-animation branch "no coverage at all" — false, `prepared-image-policy-matrix.test.ts` already covers it under the production policy and predates the merge (retraction left visible above). `landing-image-budget.ts` attributed the importer's per-image check to `stashImageMetadataAgreesWithBlob`, an identifier that exists nowhere in the repo. And it described that writer as the legacy-stash migration when `importImagesIntoLanding` has **two** production callers — the second replays inline `b64content` nodes through `materializeInlineImages`, which gates on MIME only and applies no size ceiling. The 5 MiB conclusion survives, but it rests on what past builds could have written rather than on a check in this tree, and the docblock now says so.

**The restored file arrived with a flake**, caught by running it twice: it passed, then timed out at 20 s on unchanged code. Two boundary cases built 5 MiB animated-WebP byte arrays — 57.5 s of test time for the file, close enough to the limit that parallel load decides it, which on a contended CI shard is a red with nothing wrong. Both test the *relationship* between the ceilings, which is scale-free, so they now run on a local policy of the same shape two orders of magnitude smaller: **57.5 s → 0.25 s**, stable over three runs.

Local validation on the final tree: compile, lint and format green (uncached, six projects), full gui-app suite **2369/2369 files, 30392 passed, 0 failed**, requested files equal to reported files (no worker death), and a tree fingerprint identical before and after with zero files written during the run.

### CODE REVIEW

File: clients/gui-app/src/hooks/chats/use-initial-chat-handoff-driver.ts [hunk 1]
```diff
@@ -108,10 +138,129 @@ export function useInitialChatHandoffDriver(
...
+    getDraftBlobBridgeSupported,
+    hostId,
+    key,
+    nodeId,
+    nothingToDecide,
+  ]);
+  if (nothingToDecide) return content;
+  // Key-matched so a handoff replaced while its reads were in flight (a second
+  // create in the same epic) can never send the previous one's message.
+  return resolved !== null && resolved.key === key ? resolved.content : null;
```
[the remark this answers is not in the document]
Comment (hdkshingala) [reply clients/gui-app/src/hooks/chats/use-initial-chat-handoff-driver.ts]: Confirmed and fixed in `cf957d8c1`.

`initialChatHandoffKey` is `[scope.userId ?? "user:none", scope.epicId].join(KEY_SEPARATOR)` (`initial-chat-handoff-store.ts:402-404`) and `register` writes `handoffs[key] = {...}` at that same key, so a replacement keeps it. The guard was therefore invariant across exactly the case its own comment named ("a handoff replaced while its reads were in flight (a second create in the same epic)").

The resolved state is now keyed on `handoff.clientActionId ?? handoff.key`. `RegisterInitialChatHandoffInput` declares `clientActionId: string` (required, pre-minted at submit), so it changes per registration; the `?? key` arm is only for a persisted record predating the field, where it is no weaker than the previous behaviour. The effect dependency moved with it.

Swept for siblings: this was the only identity/freshness comparison of a handoff key in the tree.

File: clients/gui-app/src/hooks/composer/use-composer-paste.ts [hunk 2]
```diff
@@ -908,17 +948,26 @@ async function hashImageAttrsFromFiles(
...
           signal,
-          () => `Reading ${file.name || "image"} timed out`,
+          trackRejected,
         );
-        const bytes = new Uint8Array(buffer);
+        // Refused outright: it cannot be made to fit, and the toast has
+        // already named it. Its reservation slot is left unsettled and the
+        // batch handle releases it with the rest.
+        if (prepared === null) return null;
+        const bytes = prepared.bytes;
```
[the remark this answers is not in the document]
Comment (hdkshingala) [reply clients/gui-app/src/hooks/composer/use-composer-paste.ts:995]: Confirmed and fixed in `cf957d8c1`, though not by the reordering — the charge is now capped instead.

The numbers check out: `LANDING_IMAGE_BUDGET_BYTES` is 64 MiB, `PREPARED_IMAGE_MAX_BYTES` is 3,932,160 (3.75 MiB), and `collectImages` admits up to `MAX_IMAGE_SOURCE_BYTES`. It also contradicted the budget module's own invariant — `LANDING_IMAGE_MAX_BYTES_PER_IMAGE` is defined as `max(PREPARED_IMAGE_MAX_BYTES, PROMPT_STASH_IMAGE_MAX_BYTES)` = 5 MiB and documented as "the largest blob any writer can land", while this site could reserve ten times that.

I capped the per-candidate charge at `LANDING_IMAGE_MAX_BYTES_PER_IMAGE` rather than preparing first. Every arm that reaches `putImage` has already passed preparation, and `prepareComposerImageBytesOrRefuse` refuses anything over `PREPARED_IMAGE_MAX_BYTES` on both the re-encode path and the fallback — so the cap is a true upper bound on what can land, and `settleStored` replaces it with the real charge as soon as the hash exists.

Keeping the order is deliberate: preparation is the expensive part, and reserving first is what stops N concurrent batches from all decoding before any of them learns there is no room. Landing can prepare first because it prepares serially anyway; hoisting preparation out of this batch's `Promise.allSettled` would also serialise the reads, on the path this change set exists to speed up.

Sweep: of the four `reserveLandingImageBudget` call sites, this was the only one charging a source size — the other three already charge a prepared `byteLength`.

File: clients/gui-app/src/lib/composer/__tests__/landing-image-budget-ceiling-covers-every-writer.test.ts [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+    // ceiling named.
+    const store = readSource("src/lib/composer/landing-image-store.ts");
+    expect(store).toContain("export async function putImage(");
+
+    const declared = new Set(WRITERS.map((writer) => writer.path));
+    for (const path of declared) {
+      expect(readSource(path), path).toContain("putImage(");
+    }
+    expect(declared.size).toBe(WRITERS.length);
+  });
```
[the remark this answers is not in the document]
Comment (hdkshingala) [reply clients/gui-app/src/lib/composer/__tests__/landing-image-budget-ceiling-covers-every-writer.test.ts:241]: You are right, and this one is mine — the test's name promised a completeness check its body never performed. Fixed in `cf957d8c1`.

It now builds the caller set from the tree (`productionSources()` walks `src`, excluding `__tests__` and `*.test.*`) and compares it to `WRITERS` **both ways**, rather than only checking that declared paths contain `putImage(`.

The predicate is the imported BINDING, not the call text: a bare `putImage(` also matches `putImageBytesAtHash(` and matches the word in a comment — a dozen files in this tree mention it without calling it. Two guards on top: a control asserting `cloud-draft-image-recovery.ts` is excluded (it writes bytes, mentions `putImage` in prose, and imports `putImageBytesAtHash` from the same module, so a sloppier predicate pulls it in), and a check that the walk really read the tree.

Ablated: adding a file that imports and calls `putImage` reds this exact assertion and names the file. The current set turns out to be complete (the same 6), so this was a latent hole rather than a live undercount.

I also wrote down why `putImageBytesAtHash` callers are out of scope for the ceiling: that function only stores bytes whose SHA-256 already matches a hash some document names, so it re-materialises a blob a writer already minted — the same argument `landing-image-move.ts` carries.

File: clients/gui-app/src/lib/composer/landing-image-budget.ts [hunk 4]
```diff
@@ -187,6 +253,14 @@ function declaredSizeByHash(
       if (!sizeByHash.has(atom.hash)) sizeByHash.set(atom.hash, atom.size ?? 0);
     }
   }
+  // First writer wins throughout, including here, so a draft's own declaration
+  // is never displaced by a holder's. Both are the same node's `size` in
+  // practice; the order only decides which copy answers.
+  for (const source of extraSizeSources) {
+    for (const [hash, size] of source.declaredSizes()) {
+      if (!sizeByHash.has(hash)) sizeByHash.set(hash, size);
+    }
```
[the remark this answers is not in the document]
Comment (hdkshingala) [reply clients/gui-app/src/lib/composer/landing-image-budget.ts:295]: Verified the mechanism and I'm leaving the behaviour as is — reasoning rather than a silent dismissal.

You're right that `putImageBytesAtHash` verifies the digest and not the length, so a declaration can understate. What bounds it is that the undercharge is transient and self-correcting: `rootByteCost` reads `measuredLandingImageSize(hash)` **first** and only falls through to the declared value for a hash this partition has never held. The moment recovery writes the bytes, the measurement takes precedence and the accounting corrects itself — the window is exactly "declared but not yet fetched".

The proposed alternative costs more than it saves. Charging the per-image ceiling for every unmeasured declared root taxes every restored draft for the whole of that window, which is the failure the sibling assertion in `landing-image-budget-ceiling-covers-every-writer.test.ts` exists to prevent ("a ceiling above every writer is also a tax"), and it is the direction that refuses pastes which would have fit. Validating the declaration against the fetched bytes would only move the correction to where the measurement already happens.

The declaration is also produced by the same local ingest that wrote the node's `size`, and the budget is soft admission control over a local IndexedDB partition, not a security boundary.

Happy to revisit if you can show a path where a declaration and its blob disagree by enough to matter after measurement lands.

File: clients/gui-app/src/lib/composer/strip-base64-image-nodes.ts [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+ * omits that paste's image from the serialized form, because its b64 node has
+ * not become a hash yet.
+ */
+import type { JsonContent } from "@traycer/protocol/common/registry";
+
+import { EMPTY_LANDING_DRAFT_CONTENT } from "@/stores/home/landing-draft-content";
+
+export function stripBase64ImageNodes(content: JsonContent): JsonContent {
+  return stripBase64ImageNode(content) ?? EMPTY_LANDING_DRAFT_CONTENT;
+}
```
[the remark this answers is not in the document]
Comment (hdkshingala) [reply clients/gui-app/src/lib/composer/strip-base64-image-nodes.ts:100]: Confirmed, and sharper than "can fail" — fixed in `cf957d8c1`.

`@tiptap/extension-document` declares `content: "block+"`, so a doc with zero children is not a degraded draft, it throws while ProseMirror builds it: the composer restoring that draft fails to mount rather than coming up empty.

It was also already reachable in a pinned test. `landing-draft-store.test.ts` had a case — "drops an attachmentGroup left empty after stripping its only b64 child" — whose fixture is exactly your scenario and which asserted `{ type: "doc", content: [] }`, with a comment describing what the walker happened to produce. That pin now expects the empty document, and says why it changed.

`stripBase64ImageNodes` returns `EMPTY_LANDING_DRAFT_CONTENT` when the stripped root is a `doc` with no children. The existing `?? EMPTY_LANDING_DRAFT_CONTENT` covered a root stripped away entirely; this covers a root that survived with nothing left in it, which is the case `??` could not see because an empty doc is not null.

File: clients/gui-app/src/stores/composer/composer-draft-store.ts [hunk 6]
```diff
@@ -507,6 +508,30 @@ export const useComposerDraftStore = create<ComposerDraftStore>()(
...
+      // prepare+hash+store job is keyed on, and the composer re-enters ingest
+      // for it on every document change - so the strip lives only here, at the
+      // localStorage seam. A hash-only node, whose bytes are durable in the
+      // composer image store, always survives, and `blobHashes` on the draft
+      // mirror's write is derived from those same hashes, which is what shrank
+      // the debounced `drafts.upsert` bodies from megabytes to a few KB.
+      //
+      // `selection` is left alone: a caret offset counts NODES, and the image
+      // node is still there in memory - only its serialized copy loses one, and
+      // the rehydrated document is what the caret is then applied to.
```
[the remark this answers is not in the document]
Comment (hdkshingala) [reply clients/gui-app/src/stores/composer/composer-draft-store.ts]: Confirmed and fixed in `cf957d8c1` — and there were three of these, not one.

The comment that was there argued the in-memory case ("the image node is still there in memory"), which is true and answers the wrong question: what the caret is applied to on the next launch is the **rehydrated** document, and that is the stripped copy.

Sweeping `stripBase64ImageNodes` call sites for "strips content, persists a selection" found two more beyond the one you flagged — `landing-draft-store.ts` in both its persist `partialize` and its desktop projection. (`initial-chat-handoff-store.ts` strips too but carries no selection.)

So the rule moved into the strip module as `stripBase64ImageNodesWithSelection`, which returns the pair, and all three seams take it — a fourth seam gets it by construction rather than by remembering. The caret is dropped rather than rebased: rebasing needs real positions, which needs the schema and a built document, inside a `partialize` on plain JSON, for a window that already loses the image itself.

Two things the fix had to get right that are worth noting: the strip still runs unconditionally and the predicate decides only the caret, because the walker also drops an already-empty `attachmentGroup` that has no b64 node to detect; and there is a control asserting the caret **survives** a draft with nothing to strip, since dropping it unconditionally would be its own regression.

File: clients/gui-app/src/lib/composer/strip-base64-image-nodes.ts [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+  selection: TSelection | null,
+): { readonly content: JsonContent; readonly selection: TSelection | null } {
+  // The strip runs unconditionally, and the predicate decides the CARET only.
+  // Short-circuiting the walk on `!containsBase64ImageNodes` looks equivalent
+  // and is not: the walker also drops an `attachmentGroup` that is already
+  // empty, which has no b64 node to detect. Skipping it there would quietly
+  // start persisting a shape these seams have always removed.
+  return {
+    content: stripBase64ImageNodes(content),
+    selection: containsBase64ImageNodes(content) ? null : selection,
```
[the remark this answers is not in the document]
Comment (hdkshingala) [reply clients/gui-app/src/lib/composer/strip-base64-image-nodes.ts]: Fixed in 4cad95a64 — though not with the proposed shape, and the reachability is narrower than the finding states, which is worth writing down.

**The logic finding is right.** Line 87 keyed the caret on `containsBase64ImageNodes` while the content half ran the walk unconditionally, and the walk drops an `attachmentGroup` that arrives *already* empty (line 120) — a change that predicate cannot see. The comment three lines above the bug already said exactly this: the round-2 fix corrected the content half and left the caret on the old predicate, so this is a half-fix of mine, not something the module always had.

**Not a second predicate.** `containsEmptyAttachmentGroups` is the same arrangement one shape further along: two hand-maintained derivations of "did the tree change", which is precisely how this one fell out of step with the walker. Instead `stripBase64ImageNode` now returns its **input** when it changes nothing, so `stripped === content` is not a proxy for "the tree is untouched" — it is that statement — and a future arm in the walker cannot escape it the way this one did. `containsBase64ImageNodes` had no other caller and is deleted; its docblock claimed to answer *"whether `stripBase64ImageNodes` would change it"*, which is the claim that stopped being true when the empty-group arm was added. Returning the input also skips an allocation per node on every persist of a draft that is not mid-paste — the saving that predicate was originally introduced to get.

**Reachability, stated precisely.** `attachment-group-extension.ts` declares `content: "imageAttachment+"`, so a live editor cannot hold a zero-child group, and the strip is itself the seam that removes one. So this is a defence-in-depth arm made self-consistent, **not** a caret corruption reachable today — the scenario in the finding needs a state the schema forbids. Still worth fixing: the arm exists, `content-recovery.ts` independently defends the same shape, and an invariant that holds only by virtue of another file's schema is the kind that stops holding quietly.

**Validation.** The module had no test file of its own despite four non-obvious arms; `__tests__/strip-base64-image-nodes.test.ts` adds ten, including the old predicate kept as a **control**. Ablating the fix fails exactly one test — the empty-group caret case — and leaves the other nine green, which is the evidence that the new case is pinned and the others are not accidentally coupled to it. Full gui-app suite green (2368/2368 files, 30378 tests), compile and lint green.

**Comment by hdkshingala** [comment]
All seven review threads are resolved and every finding has a fix in the branch, not just a reply:

- The `strip-base64-image-nodes.ts` caret finding is fixed in `4cad95a64` — though not with the proposed second predicate. `stripBase64ImageNode` now returns its *input* when it changes nothing, so `stripped === content` **is** "the tree is untouched" rather than a re-derivation that can drift from the walker, which is how the original one drifted. `containsBase64ImageNodes` had no other caller and is gone. The module had no test file despite four non-obvious arms; it has one now, with the old predicate kept as a control — ablating the fix reddens exactly the empty-attachment-group caret case and leaves the other nine green.

  One correction to the finding's framing, recorded rather than glossed: `attachment-group-extension.ts` declares `content: "imageAttachment+"`, so a live editor cannot hold a zero-child group. This is a defence-in-depth arm made self-consistent, not a caret corruption reachable today.

- A follow-up cold review of the merge found coverage deleted without a conflict — `prompt-stash-image-preparation-animation.test.ts` was removed by main with the stash plane while the module it tested was *rescued* under a new name. Thirteen tests re-homed in `ca748ca67`, including the verbatim-animation decode-validate invariant that had no test at all.

The standing `CHANGES_REQUESTED` is from `cf957d8c1`, three heads ago. Current head `90f74a2d5` is 38/38 green with zero open threads.

@coderabbitai approve
